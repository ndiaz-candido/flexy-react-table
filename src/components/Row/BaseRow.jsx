import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  LeftSection,
  RightSection,
  ScrollableSection,
} from './sections';

const defaultRowStyle = {
  display: 'flex',
  overflow: 'hidden',
  minHeight: '50px',
};
const stickyStyle = {
  position: 'sticky',
  top: '0',
  zIndex: '1000',
};
const cellDefaultStyles = width => ({
  width: `${width}px`,
  flex: `0 0 ${width}px`,
});

class BaseRow extends Component {
  static DEFAULT_COLUMN_WIDTH = 50;
  static HEADER_ROW_ID = 0;
  static NOT_SPECIFY_COLUMN_VALUE = 'Not Column';

  state = {
    leftStickySection: [],
    rightStickySection: [],
    scrollableSection: [],
  };

  static getDerivedStateFromProps({ columns, rowData, columnClassName }) {
    const leftStickySection = [];
    const rightStickySection = [];
    const scrollableSection = [];

    columns.forEach(column => {
      const columnWidth = column.width || BaseRow.DEFAULT_COLUMN_WIDTH;
      const columnDataKey = column.dataKey || '';
      const cellContent = rowData[columnDataKey] || '';
      const styles = cellDefaultStyles(columnWidth);
      const cell = (
        <div className={columnClassName} style={styles}>
          {cellContent}
        </div>
      );

      switch (column.stickyAlign) {
        case 'left':
          leftStickySection.push(cell);
          break;
        case 'right':
          rightStickySection.push(cell);
          break;
        default:
          scrollableSection.push(cell);
          break;
      }
    });

    return {
      leftStickySection,
      rightStickySection,
      scrollableSection,
    };
  }

  render() {
    const { className, isSticky, onScroll, rowId, scrollLeft } = this.props;
    const {
      leftStickySection,
      rightStickySection,
      scrollableSection,
    } = this.state;
    const stickyStyleProps = isSticky ? stickyStyle : {};
    const styles = {
      ...defaultRowStyle,
      ...stickyStyleProps,
    };

    return (
      <div className={className} style={styles}>
        <LeftSection cells={leftStickySection} />
        <ScrollableSection
          rowId={rowId}
          cells={scrollableSection}
          onScroll={onScroll}
          scrollLeft={scrollLeft}
        />
        <RightSection cells={rightStickySection} />
      </div>
    );
  }
}

BaseRow.propTypes = {
  columns: PropTypes.array,
  className: PropTypes.any,
  columnClassName: PropTypes.any,
  isSticky: PropTypes.bool,
  isHeader: PropTypes.bool,
  onScroll: PropTypes.func,
  rowId: PropTypes.number,
  rowData: PropTypes.object,
  scrollLeft: PropTypes.number,
};

BaseRow.defaultProps = {
  columns: [],
  className: '',
  columnClassName: '',
  isSticky: false,
  isHeader: false,
  onScroll: () => {},
  rowId: -1,
  rowData: {},
  scrollLeft: 0,
};

export default BaseRow;
