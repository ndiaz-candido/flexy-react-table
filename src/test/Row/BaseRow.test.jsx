import React from 'react';
import { shallow } from 'enzyme';
import BaseRow from '../../components/Row/BaseRow';

describe('BaseRow', () => {
  const columns = [
    {
      dataKey: 'a',
      label: 'a',
      stickyAlign: 'left',
    },
    {
      dataKey: 'b',
      label: 'b',
      width: '150',
    },
    {
      dataKey: 'c',
      label: 'c',
      width: '50',
      stickyAlign: 'right',
    },
  ];

  test('should render its default content', () => {
    const BaseRowComponent = shallow(<BaseRow />);

    expect(BaseRowComponent).toMatchSnapshot();
  });

  test('should render a cell per column', () => {
    const rowData = {
      a: 'Row A / Column A',
      b: 'Row A / Column B',
      c: 'Row A / Column C',
    };
    const BaseRowComponent = shallow(
      <BaseRow columns={columns} rowData={rowData} />
    );

    expect(BaseRowComponent).toMatchSnapshot();
  });

  test('should render an empty cell if no data key is provided', () => {
    const column = {
      label: 'a',
    };
    const rowData = {
      a: 'Row A / Column A',
    };
    const BaseRowComponent = shallow(
      <BaseRow columns={[column]} rowData={rowData} />
    );

    expect(BaseRowComponent).toMatchSnapshot();
  });

  test('should set isSticky when isSticky is provided', () => {
    const BaseRowComponent = shallow(<BaseRow isSticky />);

    expect(BaseRowComponent).toMatchSnapshot();
  });
});
