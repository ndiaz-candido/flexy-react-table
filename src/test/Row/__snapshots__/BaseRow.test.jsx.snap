// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseRow should render a cell per column 1`] = `
<div
  className=""
  style={
    Object {
      "display": "flex",
      "minHeight": "50px",
      "overflow": "hidden",
    }
  }
>
  <LeftSection
    cells={
      Array [
        <div
          className=""
          style={
            Object {
              "flex": "0 0 50px",
              "width": "50px",
            }
          }
        >
          Row A / Column A
        </div>,
      ]
    }
  />
  <ScrollableSection
    cells={
      Array [
        <div
          className=""
          style={
            Object {
              "flex": "0 0 150px",
              "width": "150px",
            }
          }
        >
          Row A / Column B
        </div>,
      ]
    }
    onScroll={[Function]}
    rowId={-1}
    scrollLeft={0}
  />
  <RightSection
    cells={
      Array [
        <div
          className=""
          style={
            Object {
              "flex": "0 0 50px",
              "width": "50px",
            }
          }
        >
          Row A / Column C
        </div>,
      ]
    }
  />
</div>
`;

exports[`BaseRow should render an empty cell if no data key is provided 1`] = `
<div
  className=""
  style={
    Object {
      "display": "flex",
      "minHeight": "50px",
      "overflow": "hidden",
    }
  }
>
  <LeftSection
    cells={Array []}
  />
  <ScrollableSection
    cells={
      Array [
        <div
          className=""
          style={
            Object {
              "flex": "0 0 50px",
              "width": "50px",
            }
          }
        >
          
        </div>,
      ]
    }
    onScroll={[Function]}
    rowId={-1}
    scrollLeft={0}
  />
  <RightSection
    cells={Array []}
  />
</div>
`;

exports[`BaseRow should render its default content 1`] = `
<div
  className=""
  style={
    Object {
      "display": "flex",
      "minHeight": "50px",
      "overflow": "hidden",
    }
  }
>
  <LeftSection
    cells={Array []}
  />
  <ScrollableSection
    cells={Array []}
    onScroll={[Function]}
    rowId={-1}
    scrollLeft={0}
  />
  <RightSection
    cells={Array []}
  />
</div>
`;

exports[`BaseRow should set isSticky when isSticky is provided 1`] = `
<div
  className=""
  style={
    Object {
      "display": "flex",
      "minHeight": "50px",
      "overflow": "hidden",
      "position": "sticky",
      "top": "0",
      "zIndex": "1000",
    }
  }
>
  <LeftSection
    cells={Array []}
  />
  <ScrollableSection
    cells={Array []}
    onScroll={[Function]}
    rowId={-1}
    scrollLeft={0}
  />
  <RightSection
    cells={Array []}
  />
</div>
`;
